File: jbake-core/src/test/java/org/jbake/render/ServiceLoaderTest.java
package org.jbake.render;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;
import java.util.ServiceLoader;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertTrue;

class ServiceLoaderTest {

    @Test
    void testLoadRenderer() throws Exception {

        URL serviceDescription = ClassLoader.getSystemClassLoader().getResource("META-INF/services/org.jbake.render.RenderingTool");
        File services = new File(serviceDescription.toURI());
        assertTrue(services.exists(), "Service definitions File exists");

        FileReader fileReader = new FileReader(services);
        BufferedReader reader = new BufferedReader(fileReader);

        String serviceProvider;
        List<String> renderingToolClasses = new ArrayList<String>();

        for (RenderingTool tool : ServiceLoader.load(RenderingTool.class)) {
            renderingToolClasses.add(tool.getClass().getName());
        }

        while ((serviceProvider = reader.readLine()) != null) {
            assertTrue(renderingToolClasses.contains(serviceProvider), "Rendering tool " + serviceProvider + " loaded");
        }
    }
}

File: jbake-core/src/test/java/org/jbake/render/ServiceLoaderTest.kt
package org.jbake.render

import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe
import java.io.BufferedReader
import java.io.File
import java.io.FileReader
import java.util.*

class ServiceLoaderTest : StringSpec({

    "testLoadRenderer" {
        val serviceDescription =
            ClassLoader.getSystemClassLoader().getResource("META-INF/services/org.jbake.render.RenderingTool")
        val services = File(serviceDescription!!.toURI())
        services.exists() shouldBe true

        val fileReader = FileReader(services)
        val reader = BufferedReader(fileReader)

        var serviceProvider: String?
        val renderingToolClasses: MutableList<String> = ArrayList<String>()

        for (tool in ServiceLoader.load(RenderingTool::class.java)) {
            renderingToolClasses.add(tool.javaClass.getName())
        }

        while ((reader.readLine().also { serviceProvider = it }) != null) {
            renderingToolClasses.contains(serviceProvider) shouldBe true
        }
    }
})
